emit on current client change

Dhoot opened the pull request:
This is required to support AIM specific Agent Scheduler, because we are already maintaining current client in DDS after this SB will won't use another DDS to do the same.

Commit:
emit on current client change

Commit:
emit on current client change

tylerbutler committed:
Remove Hugo version from CI (#7665)

It's not needed since we install our own version of Hugo at npm install time.

tylerbutler committed:
Make API rollup more correct and complete (#7652)

The API rollup was missing the test-client-utils package, and wasn't rewriting all the references to imports.

vladsud commented on packages/runtime/agent-scheduler/src/scheduler.ts:
```diff
@@ -217,6 +217,9 @@ class AgentScheduler extends TypedEventEmitter<IAgentSchedulerEvents> implements
         // eslint-disable-next-line @typescript-eslint/no-misused-promises
         this.consensusRegisterCollection.on("atomicChanged", async (key: string, currentClient: string | null) => {
+            // emit on current client change
+            this.emit((!currentClient) ? "taskPicked" : "taskLost", key);
```
There is already "picked" & "released" events, this seems to duplicate them. Why?

Dhoot replied:
My bad, I didn't realise that it's already there

Dhoot replied:
Hi @vladsud,

I check that again but  "picked" & "released" not emitted to all connected client it just emit with condition currentClient === this.clientId

but we want this to be emitted for all connected client.

tylerbutler committed:
Add basic package documentation for external API packages (#7669)

tyler-cai-microsoft committed:
Remove delta manager export (#7634)

Removes the `deltaManager.ts` file from export as well as privatizing `resumeInterval` in the `Container` class. 

Co-authored-by: Tyler Butler <tyler@tylerbutler.com>

tylerbutler committed:
Update code owners for public API reports (#7667)

tylerbutler committed:
Layer-check: add support for specifying an output path for the markdown (#7668)

pradeepvairamani committed:
Metric to track connect doc (#7676)

* Do not give SummaryWrite scope to clients that are not summarizers

* Connect document metrics

* Fix tests

vladsud committed:
fluid:telemetry:OdspDriver:DeltaConnection:GetOpsTooMany events in prod telemetry (#7643)

Closes #7602

These events should be error events.
Also collect a bit more data on them.
And address off-by-1 error.

Based on looking at telemetry, some of these errors might be caused by a bug that caused us to fetch ops from seq=1, and PUSH being slow to respond to these events. This bug has been fixed already.

vladsud committed:
Proper workflow to allow full offline support in presence of cached snapshot. (#7642)

Closes #7639
Before this change, we fail in offline if network request happens to return with failure sooner than cache comes back with results.

tylerbutler committed:
Add release notes for 0.48.3 and 0.48.4 (#7672)

hedasilv committed:
Preventing Lumberjack from throwing errors (#7675)

* Preventing Lumberjack from throwing errors

* Fixing UTs

* removing extra lines

* fixing ut name

GaryWilber committed:
Expose some RdKafka specific functionality in IProducer & IConsumer interfaces (#7683)

hedasilv committed:
Adding RunService metric and instrumenting rest of services-shared with Lumberjack (#7653)

* Adding RunService metric and instrumenting rest of services-shared with Lumberjacl

* Adding wait to allow telemetry date to be emitter before exiting node process

* fixing spacing

* Adding optional parameter to override default wait time before exit

scarlettjlee committed:
Use updated type of encoding (Part 3) (#7662)

* Update client packages with new dependency versions for encoding type

* Update server package versions

jatgarg committed:
fix running e2e tests for release branches (#7685)

jatgarg committed:
Policy fix check (#7686)

evaliyev committed:
chore: convert @fluid-experimental/property-common to TS [ 2 /2 ] (#6821)

scarlettjlee committed:
Breaking change notes (#7687)

jatgarg committed:
Update @fluidframework/gitresources package (#7689)

vladsud committed:
Properly log websocket errors (#7670)

We were not logging WebSocket error when socket.io failed to connect to websocket

DLehenbauer committed:
Property DDS: Test discoverability (#7324) (#7620)

jatgarg committed:
Update build-tools version (#7703)

jatgarg committed:
[bump] package version to 0.50.0 for development after client release (#7706)

alpaix committed:
Use container ID in webpack loader (#7682)

* Use container ID in webpack loader

* Add comments

pradeepvairamani committed:
Add lumberjack logs for scribe (#7696)

* Do not give SummaryWrite scope to clients that are not summarizers

* Scribe lumberjack logs

tylerbutler committed:
Redirect /playground to homepage (#7723)

jatgarg committed:
Fix detection of pre release deps in build pipeline (#7729)

jatgarg committed:
Fix package order in which they are built (#7733)

sonalivdeshpande committed:
Tinylicious Documentation (#7608)

Tinylicious Documentation

TriDucNgo committed:
riddler API to support soft/hard deletion (#7701)

* riddler API to support soft/hard deletion

* feedback update (1): use "Include-Disabled" header

* feedback update (2): scheduledDeletionTime & route

evaliyev committed:
Convert @fluid-experimental/property-properties to ES6 (#7568)

curtisman committed:
Fix bump prerelease version after release that doesn't ends with -0 (#7742)

znewton committed:
r11s-driver: share snapshot and blob cache with summarizer (#7741)

jatgarg committed:
Update build tools deps (#7746)

hedasilv committed:
Enabling Redis caching in Historian for WholeSummaries (#7691)

* Enabling Redis caching in Historian for WholeSummaries

* Addressing comments and disabling browser caching for GET summary calls

* Updating deleteIfExists to delete only, updating return types and method signatures

vladsud committed:
Improvements to setAutoReconnect flow (#7699)

Closes #7651. Please see issue for more details.

TriDucNgo committed:
use scheduledDeletionTIme (#7744)

anthony-murphy committed:
Remove per op regenerate log entry in merge tree (#7714)

The merge tree has a noisy log line that will log per-op on reconnect op re-submission. This change removes the noisy log,

fixes #7695
Co-authored-by: Tony Murphy <anthonm@microsoft.com>

pradeepvairamani committed:
Add lumberjack logs for deli (#7738)

* Do not give SummaryWrite scope to clients that are not summarizers

* Deli lumberjack logs

tyler-cai-microsoft committed:
Update monaco to use sequenceDelta (#7727)

* Remove uses of "op" and replace them with uses of "sequenceDelta" in monaco

* Fix documentation error

tylerbutler committed:
Add release notes for 0.48.5 (#7750)

Commit:
updating to one event only taskAssigned

Commit:
updating to one event only taskAssigned

tanviraumi committed:
Handle .app node when whole summary is enabled. (#7731)

znewton committed:
upgrade service packages to latest in t9s (#7760)

scottn12 committed:
add tests for AzureClient (#7638)

* add tests for AzureClient

* move all env handling logic to client factory

* integrate w/ azure, trying to replicate shared map issue

* azureclient tests working with live azure service

* add back in accidentally removed dev dependency

* use .wait() to make sure map value is ready

* sort package.json

* change variable names

* use consistent .container syntax

Co-authored-by: Helio Liu <59622401+heliocliu@users.noreply.github.com>

* disable console.error during one test to prevent stack output during testing for expected error

* Set up CI with Azure Pipelines

[skip ci]

* increase timeout for specific tests instead of using cmd arg

* remove accidentally added file

* add timeout cmd arg back

* update aqueduct package version

Co-authored-by: Helio Liu <59622401+heliocliu@users.noreply.github.com>

curtisman committed:
Enable ODSP test driver driveIdCache (#7759)

The driveIdCache in the ODSP test driver was never populated, so never activated. Fix that.

znewton committed:
r11s-driver: combine first getVersions and getSnapshot call (#7751)

jatgarg committed:
Remove prelease dependencies on release packages (#7758)

alpaix committed:
Refactor hosts sample to align with new ID flow (#7709)

* Refactor hosts sample to align with new ID flow

* Fix lint

znewton committed:
Enable RestLess in R11s (#7525)

alpaix committed:
Refactor e2e tests to adhere to server-side container ID (#7581)

* Refactor e2e tests to adhere to server-side container ID

* Revert protocol changes

* Fix build error

* Build fix

* ID strategy

* fix lint error

* Fix initial ID

* Fix local driver tests

* Add comments

vladsud committed:
Remove noisy errors that are not adding any new info (#7766)

znewton committed:
Enable RestLess in Historian (#7524)

curtisman committed:
Fix stress test pipeline (#7767)

zhajie committed:
Zhajie/add lumberjack logs (#7678)

* Add Lumberjack log

* Add lumberjack logs in OSS

* Preserve a inline comment as original

* Address PR comments

* Log to throttling table

* Add server-service-telemetry dependency in package.json, fix build issue

* Fixing test output by initializing Lumberjack in test files

Co-authored-by: zhajie-mac <zhajie-mac@zhajie-macs-MacBook-Pro.local>
Co-authored-by: hedasilv <hedasilv@microsoft.com>

vladsud committed:
Making sure that container errors are reported correctly on close (#7773)

Noticed by looking at telemetry and not finding errors there :(
That's a regression from #7363
We could force callers to use right API instead of overwriting category, but we already allow that for "performance" events, and the code is a bit shorter than way, so I feel supporting this flow is better.

scarlettjlee committed:
Add decode method to IFluidSerializer (#7765)

yunho-microsoft committed:
Add lumberjack logs under services-core and services-client (#7756)

* add audit log under path: FluidFramework/server/routerlicious/packages/services-core/

* address comments

Co-authored-by: Yunpeng Hou <yunho-microsoft@DESKTOP_frlucas.europe.corp.microsoft.com>

Commit:
Adding description and defination in IAgentSchedulerEvents about taskAssigned event

Commit:
Adding description and defination in IAgentSchedulerEvents about taskAssigned event

znewton committed:
r11s-driver: use refresh param on 401 retries (#7764)

TriDucNgo committed:
optional includeDisabled query param in riddler GETs (#7768)

* includeDisabled in riddler GETs

* remove get all disabled tenants route

* use request query instead of params

* use helper to parse flag

znewton committed:
Make RestLess more specific by adding an indicator in the header (#7780)

wes-carlson committed:
add blob upload to stress test (#7508)

Add blob upload to stress tests. totalBlobCount controls the total number of blobs to be uploaded and blobSize controls the size of blobs uploaded.

Attachment blobs can be added while detached with detachedBlobCount.

curtisman committed:
Reduce the amount of fluid build warnings (#7735)

`property-properties` and `property-changeset` still have warnings since linting isn't enabled yet.

Assuming that will be fix as we work thru converting those to typescript

tyler-cai-microsoft committed:
Use sequencedelta instead of op for simple operations (#7775)

* Switch to using "sequenceDelta" from "op" for simple operations

GaryWilber committed:
Close document contexts when removing idle lambdas (#7785)

vladsud committed:
Remove OpProcessingController, use LoaderContainerTracker in propertyDDS tests (#7784)

Slight changes in reconnect logic (see PRs #7753, #7393) result in failures in PropertyDDS UTs.

Looking a bit deeper, I can't easily follow intentions in OpProcessingController.
For example, OpProcessingController.process(dm1) will leave all but dm1 paused. UTs do rely on that (it's pretty clearly from some of them), in other places it feels like it's unintended result. And I do not think it correctly waits for all pending activity to be flushed (as observed in above mentioned PRs).

Given that it's deprecated and repo uses LoaderContainerTracker (explicitly or implicitly through), it's time to do the move.

Some random tests fail with this change, but are addressed by increasing timeout.
We saw this problem before with these tests - see pending #7350

vladsud committed:
Remove two types of nacks - "Nonexistent client" & "Readonly client" (#7753)

This is part of looking into issue #7137 (nacks, specifically - tight nack loops).
This change adds logic for DeltaManager to realize when it does not have "write" connection and stop sending ops on such connections. This removes "expected" nacks from the workflow. Remaining nack types are (in most cases) logical errors that needs to be raised in telemetry and looked into (10K unsummarized ops, 1Mb+ ops, etc.).

Removing "expected" nacks from the workflows allows us to make next step - apply some kind of workflow that breaks tight infinite loop of remaining nacks types. For example, we may add delays, or better - close container if we see number of nacks in a row (this is not part of this change, I'll continue to use this issue to track this work).

The con of this approach - it adds more logic to already complicated DeltaManager.

One of the other pros - it makes things more consistent for other layers. I.e. if Container checks for connection mode and it's "write", now it knows for sure that it can send an op. Even more important - client knows that it should not send ops (for example, consensus / leader election DDSs) when it was downgraded to "read" and thus would stop sending ops, and allow this client to stay in "read" connection mode once it got there.

tylerbutler committed:
Mark all SparseMatrix-related exports as deprecated (#7791)

These exports have been deprecated for some time, but haven't been identified as such.

tylerbutler committed:
Remove "service" from "Azure Fluid Relay" (#7796)

vladsud committed:
Properly handling server_disconnect events (#7740)

A version ago, server_disconnect errors were reported as "Disconnect: socket.io: disconnect: socket.io: server_disconnect: AccessExpired". This is not great as we go through createErrorObject() twice due to server_disconnect flow using disconnect flow.
It also has an issue with using socket.emit() which is not great (as it's easy to make mistake here - most of emits are actually messages sent over socket, like "op").

But after recent changes to capture more info for web socket errors, these same errors show up as "Disconnect: socket.io (disconnect): [object omitted]", which means we lose data.

Fixing it by refactoring server_disconnect flow.

curtisman committed:
Enable using multiple tenant in odsp test driver (#7790)

anthony-murphy committed:
Type Validation: Support Multiple Versions with Breaking Changes (#7799)

This change updates the type validation tooling to support multiple versions in the package json broken definitions. This changes the format, so existing package json type validation was updated, and all tests were re-generated

related to #7614

jatgarg committed:
[release/0.49 > main]: Add url back in ICreateBlobResponse (#7801) (#7806)

znewton committed:
upgrade service packages in client, historian, t9s (#7810)

Commit:
Merge branch 'emit-on-current-client-change-in-agent-scheduler' of https://github.com/Dhoot/FluidFramework into emit-on-current-client-change-in-agent-scheduler
